plutus-core-interpreter/src/Language/PlutusCore/Interpreter/CekMachine.hs

```diff
@@ -167,32 +177,46 @@ applyEvaluate funVarEnv _         con fun                    arg =
             Nothing                       ->
                 throwError $ MachineException NonPrimitiveApplicationMachineError term
             Just (IterApp headName spine) -> do
-                constAppResult <- runQuote <$> applyStagedBuiltinName headName spine
+                constAppResult <- runQuoteT $ applyStagedBuiltinName headName spine
                 withVarEnv funVarEnv $ case constAppResult of
-                    ConstAppSuccess res -> returnCek con res
+                    ConstAppSuccess res -> computeCek con res
                     ConstAppFailure     -> pure EvaluationFailure
                     ConstAppStuck       -> returnCek con term
                     ConstAppError   err ->
                         throwError $ MachineException (ConstAppMachineError err) term
 
+-- TODO: we must catch exceptions here. Or maybe reflect them in the `KnownDynamicBuiltinType` class.
```

Comment:
Wouldn't this be a lot simpler if `Evaluate` reflected the non-throwing interface to the machines? Then you would just get back an `Either` and it would be straightforward.

Reply:
Yes, but I haven't tried it yet. Besides, I again think evaluators should throw and that making the CEK machine non-throwing was a mistake. It really should be the case that the CEK machine receives a term wrapped in `newtype Checked = Checked (Term TyName Name a)` and any error then in an internal one and we get rid of those `Either`s, checks and other garbage.

Reply:
Well, except it's hard to compose things that throw, as we're discovering here. I still think they shouldn't throw, but :shrug:

Reply:
Here we're discovering that it's hard to make throwing and non-throwing stuff play well together, no?

Reply:
Fixed. Now the entire built-in applications machinery runs monadically. That was surprising easy to do. And one can instantiate the monad to `Identity` (for the CK machine) or to `Either CekMachineException` (for the CEK machine). This is also a step towards the `CekT` transformer.

Reply:
Right - my point is that it's easy to compose things that don't throw, and it's easy to make things that don't throw throw at the edges of the program. But it's annoying to compose a thing that does throw and a think that doesn't.

So I claim that if we used the non-throwing version then our composition would be easier.

Reply:
Nice.